src/uwtools/rocoto.py

```diff
@@ -183,12 +183,74 @@ def _add_task_dependency(self, e: Element, config: dict) -> None:
         :param config: Configuration data for this element.
         """
         e = SubElement(e, STR.dependency)
+        for key in config.keys():
+            tag, _ = self._tag_name(key)
+            if tag in [STR.taskdep, STR.datadep, STR.timedep]:
+                self._add_task_dependency_operand(e, config=config)
+            elif tag in [STR.and_, STR.nand, STR.nor, STR.not_, STR.or_, STR.xor]:
+                self._add_task_dependency_operator(e, config=config)
+            elif tag in [STR.streq, STR.strneq]:
+                self._add_task_dependency_strequality(e, config=config)
+            # elif tag in [STR.True_, STR.False_]:
+            #    self._add_task_dependency_boolean(e, config=config)
+            else:
+                raise UWConfigError("Unhandled dependency type %s" % tag)
+
+    def _add_task_dependency_operand(self, e: Element, config: dict) -> None:
+        """
+        ???
+        """
         for key, block in config.items():
             tag, _ = self._tag_name(key)
             if tag == STR.taskdep:
                 self._set_attrs(SubElement(e, STR.taskdep), block)
+            elif tag == STR.datadep:
+                self._set_attrs(SubElement(e, STR.datadep), config={"attrs": block})
+            else:
+                self._set_attrs(SubElement(e, STR.timedep), config={"attrs": block})
+
+    def _add_task_dependency_operator(self, e: Element, config: dict) -> None:
+        """
+        ???
+        """
+        for key, block in config.items():
+            tag, _ = self._tag_name(key)
+            if tag == STR.and_:
+                self._set_attrs(SubElement(e, STR.and_), config={"attrs": block})
+            elif tag == STR.nand:
+                self._set_attrs(SubElement(e, STR.nand), config={"attrs": block})
+            elif tag == STR.nor:
+                self._set_attrs(SubElement(e, STR.nor), config={"attrs": block})
+            elif tag == STR.not_:
+                self._set_attrs(SubElement(e, STR.not_), config={"attrs": block})
+            elif tag == STR.or_:
+                self._set_attrs(SubElement(e, STR.or_), config={"attrs": block})
+            elif tag == STR.xor:
+                self._set_attrs(SubElement(e, STR.xor), config={"attrs": block})
```

Comment:
Probably a better and more clean way to do this vs. a huge if/elif/else loop. I will have to add in some logic to drill down into the next block of the config to see if an operand, operator, or other tag is present. I was thinking about doing this search by adding a dict of each and simply calling the dict. An example sketch below:

 ```
dict_operands = [datadep, taskdep, timedep]
dict_operators = [and, nand, nor, not, or, xor]
dict_other = [streq, strneq, True, False]

if tag in dict.operands:
    _add_task_dependency_operands()
```

Reply:
Something like
```
if tag in (STR.and_, STR.nand, < the others >, STR.xor):
    self._set_attrs(SubElement(e, tag), config={"attrs": block})
```
? Or `if tag in dict_operators:` if you have those already defined! Same idea as what you're considering for dispatching to `_add_task_dependency_operands()`.

Reply:
What is `dict_` in `dict_operands` et al? I guess I would expect something like
```
operands = (STR.datadep, STR.taskdep, STR.timedep)
operators = (STR.and_, STR.nand, STR.nor, STR.not_, STR.or_, STR.xor)
other = (STR.streq, STR.strneq, STR.true, STR.false)
```
But I might be missing something...

Incidentally, using tuples `(...)` instead of lists `[...]` is a nice idiom when you want to express (and actually enforce) that these collections _canot be mutated_ by code. The two data structures behave the same in most read-only contexts, but you can never accidentally mutate a tuple.

Reply:
Originally my thought was that `dict_` would be the list of items contained by that category, where category here is operator, operands, etc. But working through this logic again, a dict would not be the best way to get this point across and would lead to confusion overall. 

I also realize that this might not be necessary:
```
if tag in (STR.and_, STR.nand, < the others >, STR.xor):
    self._set_attrs(SubElement(e, tag), config={"attrs": block})
```
I currently have a messier version of this implemented and am getting back an error of
```
E               assert "{'offset': '&DEADLINE;'}" == {'offset': '&DEADLINE;'}
``` 
This is due to in `_set_attrs` we take the remainder of the config and convert it to a `str` --> `e.set(attr, str(val))`.
Maybe the better idea is if we are find one of these operators we will again call the `_add_task_dependency_operator` case which will be the recursive part. This will continue until we find a tag that ends the loop like `datadep` or `taskdep`.

Overall I think changing the idea to something like
```
operands = (STR.datadep, STR.taskdep, STR.timedep)
operators = (STR.and_, STR.nand, STR.nor, STR.not_, STR.or_, STR.xor)
other = (STR.streq, STR.strneq, STR.true, STR.false)

if tag in operands:
   # this will end the recursion
   _add_task_dependency_operands()

elif tag in operators:
    # recursion will continue
   _add_task_dependency_operands()

else: 
   # this will end the recursion
   ...
```

Reply:
Looks like you're on the right path with handling (and, crucially, ending) the recursion. Should the `elif` block call an `_operators()` method rather than `_operands()`, given that the check if on `tag in operators`? Maybe just a typo in your pseudocode...

Reply:
Yes it should be `operators`, thank you for that catch